src/passes/OptimizeInstructions.cpp

```diff
@@ -824,6 +824,25 @@ struct OptimizeInstructions
           return replaceCurrent(inner);
         }
       }
+      {
+        // eqz(eqz(x - y))  =>  x != y
+        //
+        // Note that this rule is not strictly needed, since inner eqz here fits
+        // the above pattern, and then we get the outer eqz on x == y which can
+        // then turn into x != y. However, the rule "i64(x) != 0  ==>  !!x"
```

Comment:
Where is this rule? If this rule is applied first, do we revisit this node in postorder?

Reply:
That rule is in the next diff chunk in this PR. You can also search for the text in `" "` to find it.

The problem is that we don't revisit that new code, as we'd need to backtrack. That pattern applies to the inner `eqz`, but in postorder we are moving outside, so we would not process it again.

Reply:
This is the new rule added on line 2387 and it can replace the current node with something like `eqz eqz (x - y)`. The iteration in `replaceCurrent` will visit that outer `eqz` again, but not the inner one. The inner node `eqz (x - y)` will never be visited again unless the entire pass is run again.

Reply:
(sorry, I hadn't reloaded the page and seen @kripken's response. Also writing this helped me be sure of everything myself!)